javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/instrumentation/indy/AdviceSignatureEraser.java

```diff
@@ -0,0 +1,106 @@
+import org.objectweb.asm.tree.ClassNode;
+
+/**
+ * This transformer is a workaround to make the life of instrumentation developers easier.
+ *
+ * <p>When inserting the instrumentation and calling the advice enter/exit methods, we insert an
+ * invokedynamic instruction to perform this method call. This instruction has a {@link
+ * java.lang.invoke.MethodType} associated, which normally corresponds to the method type of the
+ * called advice. This can however lead to problems, when the advice method signature contains types
+ * which are not visible to the instrumented class.
+ *
+ * <p>To prevent this, we instead associate the invokedynamic instruction with a type where all
+ * class reference are replaced with references to {@link Object}. To lookup the correct advice
+ * method, we pass the original type as string argument to the invokedynamic bootstrapping method.
+ *
+ * <p>Because bytebuddy currently doesn't support this customization, we perform the type erasure on
+ * the .class via ASM before bytebuddy parses the advice. In addition, we insert an annotation to
+ * preserve the original descriptor of the method.
+ */
+// TODO: replace this workaround when buddy natively supports altering the MethodType for
+// invokedynamic advices
+public class AdviceSignatureEraser {
+
+  private static final Pattern TYPE_REFERENCE_PATTERN = Pattern.compile("L[^;]+;");
+
+  public static final String ORIGNINAL_DESCRIPTOR_ANNOTATION_TYPE =
+      "L" + OriginalDescriptor.class.getName().replace('.', '/') + ";";
+
+  private AdviceSignatureEraser() {}
+
+  static byte[] transform(byte[] bytes) {
+    ClassReader cr = new ClassReader(bytes);
+    ClassWriter cw = new ClassWriter(cr, 0);
+    ClassNode classNode = new ClassNode();
+    cr.accept(classNode, 0);
+
+    Set<String> methodsToTransform = listAdviceMethods(classNode);
+    if (methodsToTransform.isEmpty()) {
+      return bytes;
+    }
+
+    ClassVisitor cv =
+        new ClassVisitor(AsmApi.VERSION, cw) {
+          @Override
+          public MethodVisitor visitMethod(
+              int access, String name, String descriptor, String signature, String[] exceptions) {
+            if (methodsToTransform.contains(name + descriptor)) {
+              String erased = eraseTypes(descriptor);
+              MethodVisitor visitor = super.visitMethod(access, name, erased, null, exceptions);
+              AnnotationVisitor av =
+                  visitor.visitAnnotation(ORIGNINAL_DESCRIPTOR_ANNOTATION_TYPE, false);
+              av.visit("value", descriptor);
+              av.visitEnd();
+              return visitor;
+            } else {
+              return super.visitMethod(access, name, descriptor, signature, exceptions);
+            }
+          }
+        };
+    classNode.accept(cv);
+    return cw.toByteArray();
+  }
+
+  private static String eraseTypes(String descriptor) {
+    Matcher matcher = TYPE_REFERENCE_PATTERN.matcher(descriptor);
+    StringBuffer result = new StringBuffer();
```

Comment:
```suggestion
    StringBuilder result = new StringBuilder();
```

Reply:
Unfortunately the`StringBuilder` based [Matcher methods](https://docs.oracle.com/javase%2F9%2Fdocs%2Fapi%2F%2F/java/util/regex/Matcher.html#appendReplacement-java.lang.StringBuilder-java.lang.String-) have been added in Java 9, so we can't use them here.
However, I could do some manual fiddeling with the match start/end indices if you prefer that.

Reply:
Oh, that is fine then.